### PR TITLE
Adding an e2e test to verify the example mci

### PR DESCRIPTION
**Pull request by nikhiljindal**
Ref https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/issues/48

* Adding the setup to run e2e tests.
* There is a basic create test for now, that creates the example mci and validates that IP address becomes healthy eventually.

Once this is merged, next steps for future PRs:
* Run this e2e test continuously on prow.
* Add more tests.
* Fix TODOs sprinkled all through this code.

cc @csbell @G-Harmon @madhusudancs

### CODE REVIEW

**Comment by nikhiljindal** [comment]
Example output with one cluster:
```
$ go run test/e2e/e2e.go --alsologtostderr
I1128 23:17:47.204997   25732 utils.go:58] Running command: gcloud config get-value project
I1128 23:17:48.083544   25732 basic.go:47] Creating an mci tvpmmjhfib with ip address name apbromyoet
I1128 23:17:48.083567   25732 basic.go:55] Creating app in cluster new-kube-gcp-project2_kubernetes
I1128 23:17:48.083625   25732 utils.go:58] Running command: kubectl --kubeconfig=minKubeconfig -f examples/zone-printer/app/ create --context=new-kube-gcp-project2_kubernetes
I1128 23:17:49.912047   25732 utils.go:58] Running command: gcloud compute addresses create --global apbromyoet
I1128 23:17:55.321095   25732 utils.go:58] Running command: sed -i -e s/\$ZP_KUBEMCI_IP/apbromyoet/ examples/zone-printer/ingress/nginx.yaml
I1128 23:17:55.324996   25732 utils.go:58] Running command: kubemci --ingress=examples/zone-printer/ingress/nginx.yaml --gcp-project=new-kube-gcp-project2 --kubeconfig=minKubeconfig create tvpmmjhfib
I1128 23:18:31.385290   25732 utils.go:58] Running command: kubemci get-status tvpmmjhfib --gcp-project=new-kube-gcp-project2
I1128 23:18:32.286426   25732 basic.go:95] Output from get-status: Load balancer tvpmmjhfib has IPAddress 35.190.81.146 and is spread across 1 clusters (new-kube-gcp-project2_kubernetes)
I1128 23:18:32.286580   25732 basic.go:97] IP Address: 35.190.81.146
I1128 23:18:32.309600   25732 utils.go:96] host http://35.190.81.146: GET returned http error 404 unreachable
I1128 23:19:02.330196   25732 utils.go:96] host http://35.190.81.146: GET returned http error 404 unreachable
I1128 23:19:32.330395   25732 utils.go:96] host http://35.190.81.146: GET returned http error 404 unreachable
I1128 23:20:02.348381   25732 utils.go:96] host http://35.190.81.146: GET returned http error 404 unreachable
I1128 23:20:32.350452   25732 utils.go:96] host http://35.190.81.146: GET returned http error 502 unreachable
I1128 23:21:02.349009   25732 utils.go:96] host http://35.190.81.146: GET returned http error 502 unreachable
I1128 23:21:32.360344   25732 utils.go:99] Got success response from http://35.190.81.146: <!DOCTYPE html>
<html>
<head>
<title>Welcome to the global site!</title>
</head>
<body>
<h1>Welcome to the global site! You are being served from us-central1-b</h1>
<p>Congratulations!</p>
I1128 23:21:32.633298   25732 utils.go:58] Running command: kubemci --ingress=examples/zone-printer/ingress/nginx.yaml --gcp-project=new-kube-gcp-project2 --kubeconfig=minKubeconfig delete tvpmmjhfib
I1128 23:22:01.015298   25732 utils.go:58] Running command: sed -i -e s/apbromyoet/\$ZP_KUBEMCI_IP/ examples/zone-printer/ingress/nginx.yaml
I1128 23:22:01.019219   25732 utils.go:58] Running command: gcloud compute addresses delete --global apbromyoet
I1128 23:22:08.318242   25732 basic.go:60] Deleting app from cluster new-kube-gcp-project2_kubernetes
I1128 23:22:08.318338   25732 utils.go:58] Running command: kubectl --kubeconfig=minKubeconfig -f examples/zone-printer/app/ delete --context=new-kube-gcp-project2_kubernetes
```

File: test/e2e/cases/basic.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	createArgs := append(kubemciArgs, []string{"create", lbName}...)
+	runCommand(createArgs)
+	defer func() {
+		// Delete the mci.
+		deleteArgs := append(kubemciArgs, []string{"delete", lbName}...)
+		runCommand(deleteArgs)
+	}()
+
+	// Tests
+	// TODO(nikhiljindal): Figure out why is sleep required? get-status should work immediately after create is successful.
```
Comment (G-Harmon) [inline test/e2e/cases/basic.go]: File an issue?

File: test/e2e/cases/basic.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+	"github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/kubemci/pkg/kubeutils"
+)
+
+// Creates a basic multicluster ingress and verifies that it works as expected.
+// TODO(nikhiljindal): Use ginkgo and gomega and possibly reuse k/k e2e framework.
+func RunBasicCreateTest() {
+	// Validate inputs and instantiate required objects first to catch errors early.
+	// TODO(nikhiljindal): User should be able to pass kubeconfig.
+	kubeconfig := "kubeconfig"
```
Comment (G-Harmon) [inline test/e2e/cases/basic.go]: change the variable name to kubeConfigPath? (to clarify it's a file and not the contents)

File: test/e2e/cases/utils.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	output, err := cmd.Output()
+	if err != nil {
+		err = fmt.Errorf("unexpected error in executing command %s: error: %s, output: %s", strings.Join(args, " "), err, output)
+		glog.Errorf("%s", err)
+	}
+	return strings.TrimSuffix(string(output), "\n"), err
+}
+
+// Returns the first IP address found in the given string.
+func findIP(input string) string {
```
Comment (G-Harmon) [inline test/e2e/cases/utils.go]: probably should be findIPv4

File: test/e2e/cases/utils.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+	regEx := regexp.MustCompile(regexPattern)
+	return regEx.FindString(input)
+}
+
+// Waits for the ingress paths to be reachable or returns an error if it times out.
+func waitForIngress(ip string) error {
+	timeoutClient := &http.Client{Timeout: IngressReqTimeout}
+	// TODO(nikhiljindal): Add support for https and also verify all paths rather than just the root one.
+	pollURL(fmt.Sprintf("http://%s", ip), "", LoadBalancerPollTimeout, LoadBalancerPollInterval, timeoutClient, false /* expectUnreachable */)
```
Comment (G-Harmon) [inline test/e2e/cases/utils.go]: Just use "+" instead of Sprintf?
document the "" parameter.

File: test/e2e/cases/basic.go (hunk 1 above)
Comment (nikhiljindal) [reply test/e2e/cases/basic.go]: https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/pull/84

File: test/e2e/cases/basic.go (hunk 2 above)
Comment (nikhiljindal) [reply test/e2e/cases/basic.go]: Done

File: test/e2e/cases/utils.go (hunk 3 above)
Comment (nikhiljindal) [reply test/e2e/cases/utils.go]: done

File: test/e2e/cases/utils.go (hunk 4 above)
Comment (nikhiljindal) [reply test/e2e/cases/utils.go]: Thanks for the comment!
Hadnt thought about it. Turns out that "+" is way faster: https://gist.github.com/dtjm/c6ebc86abe7515c988ec (67 ns vs 416ns)
Changed.

**Comment by nikhiljindal** [comment]
Thanks for the review @G-Harmon. Updated as per comments

**Pull request merged**